### ISSUE TITLE
SystemIcons outdated?

### ISSUE THREAD

**Issue by Rand-Random** [dotnet/runtime#67040]
Is the SystemIcons outdated? Should the SystemIcons return the shell icons?

As described here:
https://stackoverflow.com/questions/24257506/how-can-i-get-messagebox-icons-in-windows-8-1
`SHGetStockIconInfo` vs `LoadIcon`

**Comment by Rand-Random** [comment]
@dotnet/area-system-drawing

**Comment by RussKie** [comment]
> does Winforms require such icons?

I am not aware of any asks in the Windows Forms SDK for these. 
And in general the Windows Forms SDK tries to take what Windows provides.

**Comment by Rand-Random** [comment]
@RussKie
so I assume this means, that yes SystemIcons are outdated, and it would be safe to update them?

**Comment by RussKie** [comment]
Windows Forms SDK doesn't use SystemIcons at all - we don't have any of the referred interop definitions, and we call Windows MessageBox API directly.

**Comment by danmoseley** [comment]
@Rand-Random we're just calling LoadIcon. Do you have a pointer to what you propose we should do instead (or as well)?

**Comment by Rand-Random** [comment]
@danmoseley - calling SHGetStockIconInfo seems to get more up-to-date icons, so that would be my proposal 

if the switch is feasible I would also recommend extending systemicoons with all possibilities of SHGetStockIconInfo as described here: https://docs.microsoft.com/en-us/windows/win32/api/shellapi/ne-shellapi-shstockiconid

though I can’t say if switching from LoadIcon to SHGetStockIconInfo is the better approach than to update LoadIcon returning up-to-date icons

**Comment by danmoseley** [comment]
There's a lot of such icons: https://docs.microsoft.com/en-us/windows/win32/api/shellapi/ne-shellapi-shstockiconid
It would be good to hear from others whether they need such icons, and which, and whether they should replace the existing ones returned. What do those look like today?

**Comment by ViktorHofer** [comment]
Referring to @danmoseley's previous comment:

> It would be good to hear from others whether they need such icons, and which, and whether they should replace the existing ones returned. What do those look like today?

A good next step would be to discuss if these icons are required and loop some other people in as well.

**Comment by ViktorHofer** [comment]
cc @JeremyKuhne

**Comment by JeremyKuhne** [comment]
I think we should expose a new `StockIcons` to get at the set that are exposed via `SHGetStockIconInfo`. UI modernization is a tenet and this would help enable this for the WinForms runtime and WinForms apps. There isn't a good way to update the existing ones as some users will undoubtedly want the current behavior.

Something like the following:

``` C#
namespace System.Drrawing;

public static class SystemIcons
{
    /// <summary>
    ///  Gets the specified stock shell icon.
    /// </summary>
    public static Icon GetStockIcon(StockIcon stockIcon);
}

public enum StockIcons
{
    DocumentNoAssociation = 0,
    DocumentWithAssociation = 1,
    Application = 2,
    Folder = 3,
    OpenFolder = 4,
    Drive525 = 5,
    Drive35 = 6,
    DriveRemovable = 7,
    DriveFixed = 8,
    DriveNetwork = 9,
    DriveNetworkDisabled = 10,
    DriveCD = 11,
    DriveRAM = 12,
    World = 13,
    Server = 15,
    Printer = 16,
    // ...
}

```


Here is a current related WinForms runtime issue: https://github.com/dotnet/winforms/issues/7390

**Comment by Rand-Random** [comment]
@JeremyKuhne 
Sorry, for the late reply.

I would argue against 
> There isn't a good way to update the existing ones as some users will undoubtedly want the current behavior.

Because this can be said, about every time Windows decided to update icons.

Lets assume Windows did update the Icons for every operating system, we would already have 11 different Icon versions.
To pick up your argument you could argue that someone would like to have Windows XP Icons on Windows 11.

I would argue against and say that "SystemIcons" should ALWAYS return the icons of the current System, and not Icons from an older system.

If you want to support Icons from older Systems you should fully commit to the idea and have a public API where you can define which SystemIcon you want and from which System.

eg. 
```
Icon GetSystemIcons(SystemIcon systemIcon, string systemInfo);

//examples
var xpIcon = GetSystemIcons(SystemIcon.Question, "Window XP");
var sevenIcon = GetSystemIcons(SystemIcon.Question, "Window 7");

//with cross plattform in mind
var ubuntuIcon = GetSystemIcons(SystemIcon.Question, "Ubuntu 18");
```

**Comment by JeremyKuhne** [comment]
> Because this can be said, about every time Windows decided to update icons.

@Rand-Random yes, the difference here is that we never updated the existing ones. That would be the behavior change and the reason for the new API that always gets the current icons. If you use the new API it will be documented as being "current". There are also significantly more icons available, which also helps justify a new API.

@ViktorHofer two things here:

1. We should probably have the new API return cached icons as well
1. WinForms _will_ use this API once it is available

**Comment by JeremyKuhne** [comment]
Created an API proposal. https://github.com/dotnet/runtime/issues/83276. @ViktorHofer Once it is approved I'll represent in API review. Once we've moved System.Drawing to the WinForms repo I'll implement unless someone else is particularly eager. :)
